### DISCUSSION TITLE
LoRa integration

### DISCUSSION THREAD

**Discussion by EDsteve** [LIFsCode/ELOC-3.0#112] [General]
@LIFsCode 

Regarding the question:
_Does it make sense to build without LoRa in the future, or do all ELOCs simply have LoRa?
Otherwise, I would, by default, set the AU frequency configuration in the targets (esp32dev and esp32dev-ei) and create an additional target for the EU configuration._

The ELOC should be able to be used as a recorder (without LoRa) and detector (with LoRa). I made the LoRa module as an add-on-module because some people use the ELOC only as a recorder, meaning without the LoRa hardware. Is it problematic if the LoRa functionality is integrated into the code, even if LoRa isn't being used?  I'm not entirely sure what difference it makes.

**Comment by LIFsCode** [comment]
@EDsteve That's fine just wanted to know if this could be handled by compile switch or via config.

Since it seems to be needed more flexible I would control that purely via Jason config.

I suggest the following Lora config fields
- enable (bool)
- region (enum)
- send interval in s(integer)

Are there already any ideas how to bring in the encryption secrets?

1. App key: is it intended to be used with a single TTN app or with different ones? I.e. has the app key to be configureable?
2. DevEUI: Are there already any ideas how to get the unique DevEUI into each single ELOC? I suggest using the NVM where also the serial number is stored. But this is a very manual process.
3. NwkKey: Same applies here

**Comment by EDsteve** [comment]
@LIFsCode

- I assume that the "send-interval-option" is for testing purposes? Or is that for something like a daily status report? Which would make sense.
- Besides the region. There is also this subBand. Is that something we should have as well in the config?
- What about spreading factor? How does that gets handled?

```
const LoRaWANBand_t Region = AU915;
const uint8_t subBand = 2;  // For US915, change this to 2, otherwise leave on 0
```
To your questions:
1. Hard to tell right now. If its easier. It can be static for now. And we see how it goes. But if it's not too much effort. Make it configurable. 
2. Sounds good. So a DevEUI/Nwk key can be generated from the serial number during NVS build and stored in NVM? That would work for me, since every ELOC gets a unique serial number via NVS.

**Comment by LIFsCode** [comment]
@EDsteve 

Based on the comments in RadioLib I would set the the subBand based on the the region. If we require something else we can still add it to config.

The Spreading facotrs are controlled via Protocol based on data rates and are defined in the Region settings. See [RadioLib src](https://github.com/jgromes/RadioLib/blob/6622a0ac9954067304bbb3198bc4697dddeda533/src/protocols/LoRaWAN/LoRaWANBands.cpp#L70C1-L84C38)

### DevEUI:

This is kind of critical in the handling. As the DevEUI must be globally unique so we cannot make up our own random numbers based on serial number.

1. Do the Ra-01S Modules come with a predefined DevEUI? Some Module manufacturers provide it.
2. If not you must manually generate a new DevEUI via TTN Console for each device.

**Comment by EDsteve** [comment]
@LIFsCode 
Ohh... sorry. I just read your last post. Not sure how i missed that.

1. I couldn't find anything regarding a EUI for Ra-01S. So i am guessing not. Most bare LoRa modules in the lower price range don't have that it seems. We could generate our own unique EUIs, but that will cost a few thousand dollars upfront.
2. TTN only provides 50 free DevEUI per application. So we could make a new application for every 50 devices. But i have also read in TTN forum, that many people generate their own using this [EUI generator](https://www.thethingsnetwork.org/forum/t/cannot-register-end-device-because-50-used-but-we-have-only-30-devices/58597). Which of course wouldn't be a unique number, but seems okay for a few hundred devices.

It seems we have to decide between two approaches:
1. We generate our own keys for each device during flashing and add them to TTN when we want to register this device.
2. We let TTN generate the keys for us and add the keys during flashing. Which means all devices have to be registered before flashing.

Not sure which is the best approach. I would lean more towards the first approach, because it gives us more flexibility. Generating keys from a script before flashing could also safe some manual work. But i need to dig a bit deeper.
What do you think?

**Comment by LIFsCode** [comment]
@EDsteve Ah ok I missed that point that TTN devEUIs are limited to 50 for free but that makes sense.

Random generation could work. Actually I have no idea what happens if it comes to collisions.

Another option would be to use the ESP wifi MAC and extend it to devEUI64: https://www.thethingsnetwork.org/forum/t/deveui-for-non-hardware-assigned-values/2093/30

That would at least give use unique IDs within all ESP32 devices and we must not think about the generation.
But the method is deprecated, so no 100% guarantee to be collision free.

**Comment by EDsteve** [comment]
@LIFsCode 
MAC addresses sounds like a good idea. Good find. Both is fine with me. Whatever you prefer.
I have to add a number generator for AppKey and NWkKey into the [Autoflasher.py](https://github.com/LIFsCode/ELOC-3.0/wiki/Firmware-Update-&-Initial-Flashing-with-NVS#how-to-flash-firmware-and-nvs) script anyway, which should also save all numbers in a csv file. So i can also try to generate a DevEUI.
If i can make that work :)
